SolaceIO write connector

iht opened the pull request:
This is a follow-up PR to #31953, and part of the issue #31905.

This PR adds the actual writer functionality, and some additional testing, including integration testing.

This should be final PR for the SolaceIO write connector to be complete.

This PR fixes #31905.

iht commented:
assign set of reviewers

Abacn commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -1026,8 +1057,180 @@ abstract static class Builder<T> {
+      }
+
+      Preconditions.checkArgument(
+          getMaxNumOfUsedWorkers() > 0,
```
If I understand correctly, maxNumOfUsedWorkers and numberOfClientsPerWorker currently have a constant default, or supplied with a constant parameter. Is there a concern regarding the scalability and flexibility to scale?

iht replied:
Yes, the concern is related to Solace quotas. They are normally much more limited than the average number of threads and workers in a pipeline (in the order of hundreds or thousands, when quotas are much lower).

sjvanrossum replied:
Most sinks would name this `numShards`, would it make sense to do the same here and merge `maxNumOfUsedWorkers` and `numberOfClientsPerWorker` into `numShards` instead? As far as I understand from the Solace documentation the broker supports a global connection limit and a per username connection limit, but not a per "machine" (source IP?) connection limit.

If during testing of this sink the output parallelism was always set to the maximum number of workers, then I understand why that may have led to issues with connection limits, but I'm hoping that a change in wording may help avoid those issues.

In any case, if I'm reading this right, the sink will always open `numberOfClientsPerWorker` and randomly use one of those clients for a given bundle (keyed on `maxNumOfUsedWorkers`), but if a machine is assigned less shards than that then it will never utilize those connections to their fullest, which seems like a waste of open network connections.

bzablocki replied:
I changed name to `numShards`

Abacn commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SolaceMessageProducer.java:
```diff
@@ -0,0 +1,84 @@
+          return 0;
+        };
+
+    retryCallableManager.retryCallable(publish, ImmutableSet.of(JCSMPException.class));
```
Took a look at RetryCallableManager, it is based on com.google.cloud.RetryHelper. Solace isn't a GCP cloud product, but SolaceIO depends on various gcp artifacts for non-gcp usages currently (like here). We probably want to clean up these usages in the future. For example, retry management is preferred to use a FluentBackoff (in Java SDK core)

sjvanrossum replied:
There's also an existing dependency on Failsafe (https://github.com/failsafe-lib/failsafe) in Beam, which I'd probably recommend using.

iht replied:
The retry library does not pull any other Google Cloud dependency, it is a generic library that happens to be distributed by Google Cloud.

Abacn commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/data/Solace.java:
```diff
@@ -248,42 +249,49 @@ public abstract static class Record {
+          .setPriority(-1)
+          .setRedelivered(false)
+          .setTimeToLive(0)
+          .setAttachmentBytes(ByteString.empty().toByteArray());
```
Do we really need to import `org.apache.beam.vendor.grpc.v1p60p1.com.google.protobuf.ByteString` to do this?

sjvanrossum replied:
```suggestion
          .setAttachmentBytes(new byte[0]);
```

iht replied:
Fixed in upcoming commit.

Abacn commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/write/UnboundedBatchedSolaceWriter.java:
```diff
@@ -0,0 +1,177 @@
+import org.slf4j.LoggerFactory;
+
+@Internal
+public final class UnboundedBatchedSolaceWriter {
```
It's sounds strange to me to have an empty UnboundedBatchedSolaceWriter, UnboundedStreamingSolaceWriter class with just a single static class (...WriteDoFn). Can we remove the outer class here? Or just put everything in UnboundedSolaceWriter

iht replied:
This is the result of "code" evolution, in the end I ended up here with a single inner class. Let me refactor it.

iht replied:
Refactored in upcoming commit, I will resolve now this conversation.

Abacn reviewed (commented):
thanks, had a first pass and had a couple of comments most about

- dead letter queue pattern
- possibly excessive dependency usage
- minor package/class structure adjusts

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -1026,8 +1057,180 @@ abstract static class Builder<T> {
+      PCollection<Solace.Record> withGlobalWindow =
+          records.apply("Global window", Window.into(new GlobalWindows()));
+
+      PCollection<KV<Integer, Solace.Record>> withShardKeys =
+          withGlobalWindow.apply(
+              "Add shard key",
+              ParDo.of(new UnboundedSolaceWriter.AddShardKeyDoFn(getMaxNumOfUsedWorkers())));
+
+      String label =
+          getWriterType() == WriterType.STREAMING ? "Publish (streaming)" : "Publish (batched)";
+
+      PCollectionTuple solaceOutput = withShardKeys.apply(label, getWriterTransform(destinationFn));
+
```
The sharding that's applied here couples Beam's internal parallelism to the number of clients you intend to use (1 per transform instance per worker). Using GroupIntoBatches.withShardedKey() instead makes more sense, since Solace mentions their producer is thread safe for some producer patterns.

On that topic, and feel free to correct me if I'm wrong, since this transform is producing session-dependent messages the producer may only be accessed by multiple threads for direct messages.
The transform switches the producer index on a per bundle basis, but it does not exclusively claim the producer, so multiple threads can access a producer regardless of the delivery mode.
To uncomplicate the thread safety conditions of the producer you could complicate access to the producer. By putting the producer on its own thread and exposing an intermediate channel for message passing (e.g. a concurrent queue) you can decouple Beam's concurrency from the number of producers (1 per distinct configuration). The producer thread polls the channel to receive sendable work from multiple producers and sets up callbacks per the message's configuration (callbacks should be per instance, see other comment).
This carries minor overhead for session-dependent direct messages, but simplifies the rest of the setup. You could optionally use a Guava cache to expire and close producers after a period of inactivity. My remaining question here is why isn't this using session-independent messages?

Back to GroupIntoBatches, it covers all of the state and timer stuff that's happening in the writer fn and applying it would simplify the writer a great deal. It looks like it might not even need state and timers at all.

iht replied:
Yes, I think that `GroupIntoBatches` would be a more elegant solution here and it would not need a state & timers custom implementation.

About session-independent messages, that's actually what I am using :). I think my implementation for the producer does actually that. There is a concurrent map of producers, and each thread/worker polls for a producer to be used to do some work. The clients are closed automatically after inactivity (by the client libraries) and the producer map takes care of making sure the producer reconnects when needed.

sjvanrossum replied:
I've confused the docs and the classes introduced by this connector. The docs state that session-dependent message are created by calling `create*Message` on a `Producer` and session-independent messages are created by calling `create*Message` on the `JCSMPFactory`. The messages produced by this connector are created by the `MessageProducer` class of this connector. It's calling out to `JCSMPFactory`, so I see that this connector is indeed creating session-independent messages.

Going back over my comment then, the bit about `GroupIntoBatches` is still relevant although likely best addressed in a follow-up PR. As an improvement to this PR I think we can still remove the state variables that limit the DoFn's parallelism since producer instances are thread-safe and manage the connection to Solace on a separate I/O thread, so Beam's internal parallelism should have no external effect as long as the producer is shared between instances of a unique transform. Did you by chance experiment with sharing a producer per unique transform configuration or per unique connection configuration instead of the fixed pool of 100 producers? If multiple producers improve throughput, then would it make sense to allow users to configure the size of the producer pool?

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -1026,8 +1057,180 @@ abstract static class Builder<T> {
+
+      PCollectionTuple solaceOutput = withShardKeys.apply(label, getWriterTransform(destinationFn));
+
+      SolaceOutput output;
+      if (getDeliveryMode() == DeliveryMode.PERSISTENT) {
+        PCollection<Solace.PublishResult> failedPublish = solaceOutput.get(FAILED_PUBLISH_TAG);
+        PCollection<Solace.PublishResult> successfulPublish =
+            solaceOutput.get(SUCCESSFUL_PUBLISH_TAG);
+        output =
+            rewindow(
+                SolaceOutput.in(input.getPipeline(), failedPublish, successfulPublish),
+                windowingStrategy);
+      } else {
+        LOG.info(
+            String.format(
+                "Solace.Write: omitting writer output because delivery mode is %s",
+                getDeliveryMode()));
+        output = SolaceOutput.in(input.getPipeline(), null, null);
+      }
+
+      return output;
```
Do both of these need to be the same output type? PublishResult contains a boolean for success or failure, metrics and an error message. It seems to me like you would want to provide the user a PCollection<Solace.Record> for successses, PCollection<DlqErrorT> for failures and PCollection<SolaceStatsT> for metrics.
See the note about using the new DLQ pattern as well.

iht replied:
But that's not possible, unless we accumulate the processed `Solace.Record` and then we check all the responses, match each response to each incoming `Solace.Record` and put the right record in the right place. I think that approach would be too heavyweight. The replies are received via "callbacks" (polling by the client for responses actually, as implemented in the Solace client libraries) by any thread in any worker (potentiall a worker different to the one that sent the original message). So doing this would require shuffling too.

The reply from Solace only contains some metadata (id, timestamp), and it is this info the one that the connector returns, which is a much simpler process.

sjvanrossum replied:
My bad, I thought the record was propagated on the result. The writer should not be polling for any responses other than those for its own requests. Allowing any publish result to be emitted as output means the output could contain results from elements that belong to different bundles or windows. Reapplying the input windowing strategy is effectively useless if the input elements and output elements can appear in different bundles and windows.

I'm not convinced you're receiving responses to requests initiated on other machines though, that would be incredibly unhelpful behavior of a message broker and it makes distributed processing impractical. It also renders the latency metrics of this connector meaningless because the result of `System.nanoTime()` is not related to synchronized system or wall-clock time and can thus only be used to measure the duration between two results that were produced on the same machine.

iht replied:
Scratch my previous comment, you are right. The callbacks are received in the same client that sent the initial request. And I could keep the full record then to emit it when the callback arrives. 

For instance, currently we actually keep a timestamp of when the message is sent.

Let me think of this, and I will send more changes in an upcoming PR.

sjvanrossum replied:
So I did some digging and found a few undocumented JCSMP methods which are used by the new Java API to make the resource ownership flow a little more like what I was expecting.

The methods are marked as `@SolReserved` on `Jcsmpsession` and are named `createProducer`. If I understand correctly, a session owns the outbound connection to Solace and a producer uses that resource to produce messages. However, the documented `getProducer` method specifies that it will replace a previously returned producer, but I'd have to validate that.

The ownership and lifetime model that you probably want here is one where there's a shared, long-lived session per unique transform in a static variable, some shared map keyed on the transform configuration, and a producer per instance of the transform all multiplexed on the shared session. Every writer instance needs its own producer since the callbacks are passed as construction arguments to the producer, that's the only way to ensure that responses are routed to the originating requestor.

Could you check if repeated calls to `getProducer` invalidate prior return values like the docs say? If they do, then we'll likely need to use `createProducer` and verify with Solace that it's ok to call undocumented public methods on their classes.

iht replied:
The connector guarantees that if there is already a producer, it is reused. If the producer is closed, a new one is created. The connector only closes producers when a worker is destroyed.

The receiver that gets all the callbacks calls is the same for all the producers (static variable, which I actually have to change), so even if there is a new producer, as long as it is in the same worker, it will receive the same callbacks.

The static variable approach is actually buggy, two instances of the Write connector will use the same instance. I am changing it to a single instance receiving the callbacks per instance of the Write transform. But the same principle applies: all the callbacks are always received in the same object, and the producers are always reused for as long as Solace permits (and closed if the worker is destroyed).

sjvanrossum reviewed (commented):
Partially reviewed, I'll give this another look tomorrow.

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/BasicAuthJcsmpSessionService.java:
```diff
@@ -39,34 +43,43 @@
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+    public abstract Builder queueName(@Nullable String q);
+
```
This is qute, but could you rename that parameter to queueName?

iht replied:
I can :), in upcoming commit.

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/BasicAuthJcsmpSessionService.java:
```diff
@@ -39,34 +43,43 @@
-  private final String password;
-  private final String vpnName;
+@AutoValue
+public abstract class BasicAuthJcsmpSessionService extends SessionService {
```
Most AutoValue classes in Beam use get/set prefixes, is there a reason this class does not follow the general style?

iht replied:
get/set is used for data classes because schemas rely on the get/set prefix to detect fields.

Other classes, such as Beam PTransforms, don´t use get/set prefixes.

I just find it less verbose using the property names directly.

sjvanrossum replied:
SGTM, non-prefixed properties occur in ~18% of AutoValue class definitions in Beam.

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -1026,8 +1057,180 @@ abstract static class Builder<T> {
+                SolaceOutput.in(input.getPipeline(), failedPublish, successfulPublish),
+                windowingStrategy);
+      } else {
+        LOG.info(
+            String.format(
+                "Solace.Write: omitting writer output because delivery mode is %s",
+                getDeliveryMode()));
```
Logger methods support formatting using their own syntax.
```suggestion
        LOG.info(
            "Solace.Write: omitting writer output because delivery mode is {}",
            getDeliveryMode());
```

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/BasicAuthJcsmpSessionServiceFactory.java:
```diff
@@ -65,11 +64,15 @@ public abstract static class Builder {
-        password(),
-        vpnName());
+    BasicAuthJcsmpSessionService.Builder builder = BasicAuthJcsmpSessionService.builder();
+    if (queue != null) {
+      builder = builder.queueName(queue.getName());
+    }
```
Why did this change from throwing to checking?
Also, given that AutoValue implementations are immutable, deriving from a class that allows you to mutate the queue name seems to run counter to what the framework provides. SessionServiceFactory should probably be an interface declaring create and optionally define an abstract class that implements the interface (abstract of course) and declares equals and hashCode abstract if your intention is to have all implementors explicitly override those methods for value equality.

iht replied:
This class was initially used for the `Read` connector alone, which requires that `queue` property to be set, and should fail if it is not set.

The `Write` connector does not need a pre-existing `queue`. I changed the `queueName` parameter to `@Nullable` and this to a check since now the `queueName` value is optional.

Initially, there were different classes for `SessionService`-ish for both connectors, with different properties. Now we use a single class, for consistency across the `Read` and `Write` connectors.

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/MessageProducer.java:
```diff
@@ -0,0 +1,143 @@
+import org.apache.beam.sdk.io.solace.data.Solace;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+
+/**
+ * Interface for publishing messages to a Solace broker.
+ *
+ * <p>Implementations of this interface are responsible for managing the connection to the broker
```
This isn't an interface, was it supposed to be?

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/MessageProducer.java:
```diff
@@ -0,0 +1,143 @@
+      SerializableFunction<Solace.Record, Destination> destinationFn,
+      DeliveryMode deliveryMode);
+
+  /**
+   * Create a {@link BytesXMLMessage} to be published in Solace.
+   *
+   * @param record The record to be published.
+   * @param useCorrelationKeyLatency Whether to use a complex key for tracking latency.
+   * @param deliveryMode The {@link DeliveryMode} used to publish the message.
+   * @return A {@link BytesXMLMessage} that can be sent to Solace "as is".
+   */
+  public static BytesXMLMessage createBytesXMLMessage(
+      Solace.Record record, boolean useCorrelationKeyLatency, DeliveryMode deliveryMode) {
+    JCSMPFactory jcsmpFactory = JCSMPFactory.onlyInstance();
+    BytesXMLMessage msg = jcsmpFactory.createBytesXMLMessage();
+    byte[] payload = record.getPayload();
+    msg.writeBytes(payload);
+
+    Long senderTimestamp = record.getSenderTimestamp();
+    if (senderTimestamp == null) {
+      senderTimestamp = System.currentTimeMillis();
+    }
+    msg.setSenderTimestamp(senderTimestamp);
+    msg.setDeliveryMode(deliveryMode);
+    if (useCorrelationKeyLatency) {
+      Solace.CorrelationKey key =
+          Solace.CorrelationKey.builder()
+              .setMessageId(record.getMessageId())
+              .setPublishMonotonicMillis(TimeUnit.NANOSECONDS.toMillis(System.nanoTime()))
+              .build();
+      msg.setCorrelationKey(key);
+    } else {
+      // Use only a string as correlation key
+      msg.setCorrelationKey(record.getMessageId());
+    }
+    msg.setApplicationMessageId(record.getMessageId());
+    return msg;
+  }
```
This can be moved to a utility class since it doesn't directly depend on anything else defined in this class.

iht replied:
Got it. I will submit some other changes before, I leave this comment open for future commits.

iht replied:
Done in upcoming commit

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/MessageProducer.java:
```diff
@@ -0,0 +1,143 @@
+
+  /**
+   * Create a {@link JCSMPSendMultipleEntry} array to be published in Solace. This can be used with
+   * `sendMultiple` to send all the messages in a single API call.
+   *
+   * <p>The size of the list cannot be larger than 50 messages. This is a hard limit enforced by the
+   * Solace API.
+   *
+   * @param records A {@link List} of records to be published
+   * @param useCorrelationKeyLatency Whether to use a complex key for tracking latency.
+   * @param destinationFn A function that maps every record to its destination.
+   * @param deliveryMode The {@link DeliveryMode} used to publish the message.
+   * @return A {@link JCSMPSendMultipleEntry} array that can be sent to Solace "as is".
+   */
+  public static JCSMPSendMultipleEntry[] createJCSMPSendMultipleEntry(
+      List<Solace.Record> records,
+      boolean useCorrelationKeyLatency,
+      SerializableFunction<Solace.Record, Destination> destinationFn,
+      DeliveryMode deliveryMode) {
+    if (records.size() > SOLACE_BATCH_LIMIT) {
+      throw new RuntimeException(
+          String.format(
+              "SolaceIO.Write: Trying to create a batch of %d, but Solace supports a"
+                  + " maximum of %d. The batch will likely be rejected by Solace.",
+              records.size(), SOLACE_BATCH_LIMIT));
+    }
+
+    JCSMPSendMultipleEntry[] entries = new JCSMPSendMultipleEntry[records.size()];
+    for (int i = 0; i < records.size(); i++) {
+      Solace.Record record = records.get(i);
+      JCSMPSendMultipleEntry entry =
+          JCSMPFactory.onlyInstance()
+              .createSendMultipleEntry(
+                  createBytesXMLMessage(record, useCorrelationKeyLatency, deliveryMode),
+                  destinationFn.apply(record));
+      entries[i] = entry;
+    }
+
+    return entries;
+  }
```
Same here, can be moved to a utility class.

iht replied:
Moved in upcoming commit

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/PublishResultHandler.java:
```diff
@@ -0,0 +1,90 @@
+    PublishResult.Builder resultBuilder = PublishResult.builder();
+    String messageId;
+    if (key == null) {
+      messageId = "NULL";
```
Is this a possible state? If so, can't message id be made nullable? If not, would the empty string be an option here?

iht replied:
I am not sure if this could be a possible state, I was trying to be defensive here. 

Message id, as such, cannot be null, but I am not sure if the reply from Solace could have an empty key in certain situations.

I will change this to an empty string instead. Change coming in upcoming commit.

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/PublishResultHandler.java:
```diff
@@ -0,0 +1,90 @@
+      long latencyMillis = calculateLatency((Solace.CorrelationKey) key);
+      resultBuilder = resultBuilder.setLatencyMilliseconds(latencyMillis);
+    } else {
+      messageId = key.toString();
+    }
```
Is there a supertype or exhaustive list of types that could be compared to instead? Calling Object.toString() might not be appropriate if we don't know what type might be provided here.

iht replied:
`Object` :).

Notice that I am overrideing the `responseReceivedEx` method, and that `key` is an `Object`. That´s all we know. 

The documentation of Solace has this example (TL;DR: what I did above, except that they always use a single class for all the keys): https://tutorials.solace.dev/jcsmp/confirmed-delivery/

iht replied:
We know what are the types that are being sent, because we are sending them. The connector will send either strings or `CorrelationKey`, depending on the settings of the write connector. That `else` block should never happened. I tried to be defensive, maybe I should just log an error instead in that else block?

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/PublishResultHandler.java:
```diff
@@ -0,0 +1,90 @@
+    PublishResultsReceiver.addResult(publishResult);
+  }
+
+  private static long calculateLatency(Solace.CorrelationKey key) {
+    long currentMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime());
+    long publishMillis = key.getPublishMonotonicMillis();
+    return currentMillis - publishMillis;
+  }
```
Why are these nanos truncated to millis? It takes up the same amount of space in the correlation key anyhow.
If it's a presentational thing, you should probably deal with that when the metric is reported, but not before it's calculated.

iht replied:
Done in upcoming commit.

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SessionService.java:
```diff
@@ -195,42 +221,49 @@ private static JCSMPProperties overrideConnectorProperties(
-      props.setProperty(JCSMPProperties.PUB_ACK_WINDOW_SIZE, BATCHED_PUB_ACK_WINDOW);
-    } else {
+
+    } else if (mode == SolaceIO.SubmissionMode.LOWER_LATENCY) {
       // Send from the same thread where the produced is being called. This offers the lowest
```
Do you really want to remodel the properties of Solace's API in this connector using different names? Can't this connector expose a `.withProperties(java.util.Properties)` and convert that using `JCSMPProperties.fromProperties(java.util.Properties)`?

iht replied:
It is not a single property, it is a combination of properties that work together. In this case, it is actually two properties. I guess I could add another mode that "go fully custom", where you control all the properties. The connector actually exposes all the properties, and you can set them to whatever value, but for two specific properties (publication ack window and message callback on reactor) it makes a "sensible" choice for you. The performance of the writer is heavily impacted by those values.

iht replied:
I have added a new method that lets the user to "go fully custom". Please notice that when you create a custom `SessionService` you have full control over all the properties. With this new submission method, the connector will not override any property. So now there are three methods: high throughput (overrides 2 properties), low latency (overrides same 2 properties with different values), custom (user properties are left intact). I will resolve now this comment, as I will be submitting a new commit with those changes soon.

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/write/UnboundedBatchedSolaceWriter.java:
```diff
@@ -0,0 +1,177 @@
+   * UnboundedStreamingSolaceWriter.WriterDoFn} but also higher latency.
+   */
+  @Internal
+  public static class WriterDoFn extends UnboundedSolaceWriter.AbstractWriterDoFn {
```
See notes in SolaceIO, I believe this can be simplified using GroupIntoBatches.

iht replied:
Noted for a future PR :).

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/write/UnboundedSolaceWriter.java:
```diff
@@ -0,0 +1,394 @@
+import org.slf4j.LoggerFactory;
+
+@Internal
+public final class UnboundedSolaceWriter {
```
Can you break this out into three separate files? This is outer class is effectively a package.

iht replied:
Yes, there is a similar comment in one of the base classes. I am refactoring it to have a single outer class, since there is only a single inner class.

iht replied:
Commit ready, the two aux DoFns are moved to their own files. This file does not use inner static classes anymore, I will be resolving now this comment.

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/write/UnboundedStreamingSolaceWriter.java:
```diff
@@ -0,0 +1,144 @@
+import org.slf4j.LoggerFactory;
+
+@Internal
+public final class UnboundedStreamingSolaceWriter {
```
Can you remove this outer class and simply give WriterDoFn a more descriptive name?

iht replied:
Yes, refactoring this, see other comments.

I will update and resolve the comments when the commit is ready.

iht replied:
Commit ready, I will resolve now this comment.

iht commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -1026,8 +1057,180 @@ abstract static class Builder<T> {
+                  MapElements.into(TypeDescriptor.of(Solace.Record.class))
+                      .via(checkNotNull(getFormatFunction())));
+
+      // Store the current window used by the input
+      PCollection<Solace.PublishResult> captureWindow =
+          records.apply(
+              "Capture window", ParDo.of(new UnboundedSolaceWriter.RecordToPublishResultDoFn()));
+
+      @SuppressWarnings("unchecked")
+      WindowingStrategy<Solace.PublishResult, BoundedWindow> windowingStrategy =
+          (WindowingStrategy<Solace.PublishResult, BoundedWindow>)
+              captureWindow.getWindowingStrategy();
+
```
Some connectors do. "Transplanting" should work because what I do here is grabbing the settings of the input window, and apply the same window (window with same settings) to the output. I need to apply a global window "internally" in the connector to control for parallelization, I think it was just nice to "respect" the input window and reapply another window with the same settings to the output, instead of producing the output with the global window that is applied internally (leaking an implementation detail to the user).

bzablocki replied:
I think the issue here is that the `publishSingleMessage()` or `publishBatch()` don't return published results immediately, they return a status through an async callback. The `rewindow` operation (in the `SolaceIO.Write#expand()`) is applied on a collection of asynchronously collected message responses, which means that they might not belong to the same window, as the original message sent. Therefore, I would also incline to not 'transplant' the window, but to convert it and output as a global.

bzablocki replied:
After discussing with @stankiewicz we concluded that we should transplant the window but we should also make sure that each bundle waits for a publish response for the messages it originally published.

bzablocki replied:
Scratch that, we've decided to output the results to a GlobalWindow. This change is necessary because the output PCollection contains asynchronously received acknowledgments.  Since it's not practical to trace back to the original message's timestamp, we can't apply the same windowing strategy.

iht commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -1026,8 +1057,180 @@ abstract static class Builder<T> {
+    /**
+     * Called before running the Pipeline to verify this transform is fully and correctly specified.
+     */
+    private void validateWriteTransform(boolean usingSolaceRecords) {
```
Changed in upcoming PR

iht replied:
Actually, scratch that, how can I make this method static if I need to access non static members in the body of the fucntion?

sjvanrossum replied:
I probably should have anchored this to the entire method or an example. I meant to say that the static import of a method from Guava's `Preconditions` is generally preferred since `checkArgument` reads more fluently than `Preconditions.checkArgument` when those called in the body of a method.

iht replied:
Thanks for the clarification, changed in upcoming commit.

iht commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -961,6 +977,21 @@ public Write<T> withWriterType(WriterType writerType) {
       return toBuilder().setWriterType(writerType).build();
     }
 
+    /**
+     * Set the format function for your custom data type, and/or for dynamic destinations.
+     *
+     * <p>If you are using a custom data class, this function should return a {@link Solace.Record}
+     * corresponding to your custom data class instance.
+     *
+     * <p>If you are using this formatting function with dynamic destinations, you must ensure that
+     * you set the right value in the destination value of the {@link Solace.Record} messages.
+     *
+     * <p>In any other case, this format function is optional.
+     */
+    public Write<T> withFormatFunction(SerializableFunction<T, Solace.Record> formatFunction) {
+      return toBuilder().setFormatFunction(formatFunction).build();
+    }
```
You are right, it is a leftover from previous changes. Removing it in upcoming commit.

iht commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -1026,8 +1057,180 @@ abstract static class Builder<T> {
     public SolaceOutput expand(PCollection<T> input) {
-      // TODO: will be sent in upcoming PR
-      return SolaceOutput.in(input.getPipeline(), null, null);
+      Class<? super T> pcollClass = checkNotNull(input.getTypeDescriptor()).getRawType();
+      boolean usingSolaceRecord =
+          pcollClass
+                  .getTypeName()
+                  .equals("org.apache.beam.sdk.io.solace.data.AutoValue_Solace_Record")
+              || pcollClass.isAssignableFrom(Solace.Record.class);
```
Changed, thanks for the hint about type descriptors, I did not know they could be used for this kind of checks.

iht commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -1026,8 +1057,180 @@ abstract static class Builder<T> {
+          pcollClass
+                  .getTypeName()
+                  .equals("org.apache.beam.sdk.io.solace.data.AutoValue_Solace_Record")
+              || pcollClass.isAssignableFrom(Solace.Record.class);
```
Yes, I made the check in the opposite direction, it is now fixed and changed to use type descriptors instead. I will resolve this comment now, the change will be in an upcoming commit.

iht commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -1026,8 +1057,180 @@ abstract static class Builder<T> {
+     * Called before running the Pipeline to verify this transform is fully and correctly specified.
+     */
+    private void validateWriteTransform(boolean usingSolaceRecords) {
+      if (!usingSolaceRecords) {
+        Preconditions.checkArgument(
+            getFormatFunction() != null,
+            "SolaceIO.Write: If you are not using Solace.Record as the input type, you"
+                + " must set a format function using withFormatFunction().");
+      }
```
Changed in upcoming commit.

iht commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/MessageReceiver.java:
```diff
@@ -19,14 +19,15 @@
  * and for receiving messages from the broker.
  */
-public interface MessageReceiver {
+public interface MessageReceiver extends Serializable {
```
That code isn't mine :). It should probably not be serializable, as the receiver is created in the worker as needed, and they should never move across workers (if it needs be, it should re-connect or be re-created).

@bzablocki FYI

bzablocki replied:
Thanks, let me remove it and run some tests

iht replied:
Resolving this comment now. Fixed in commit e545ec977a6d1942590cd16663824fb82cd0e538

(pushed on behalf of @bzablocki)

iht commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SolaceMessageProducer.java:
```diff
@@ -0,0 +1,84 @@
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableSet;
+
+@Internal
+public class SolaceMessageProducer extends MessageProducer {
```
Yes, for testing purposes for instance (mock message producers).

iht commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/PublishResultHandler.java:
```diff
@@ -0,0 +1,90 @@
+    processKey(key, true, Optional.empty());
+  }
+
+  private void processKey(Object key, boolean isPublished, Optional<JCSMPException> cause) {
```
What type is null? My Scala past does not let me use "null", but I have changed it anyways, I get paid now to do Java :P. Changed in upcoming commit.

iht commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/write/PublishResultsReceiver.java:
```diff
@@ -0,0 +1,43 @@
+ * connector.
+ */
+@Internal
+public final class PublishResultsReceiver {
```
You are right, this is actually a bug. To fix it, there should be a receiver per instance of the transform. I am working on fixing it.

bzablocki replied:
Fixed in https://github.com/apache/beam/pull/32060/commits/976e89e44fe687c2e16d14ff7f8751caa1ff8c19 (commit message "[Make the PublishResult class for handling callbacks non-static to handle pipelines with multiple write transforms.](https://github.com/apache/beam/pull/32060/commits/976e89e44fe687c2e16d14ff7f8751caa1ff8c19)" in case the hash changes with a force push)

sjvanrossum replied:
SGTM, is there any need for this class at all if it's just a wrapper around a queue?

iht commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/write/UnboundedBatchedSolaceWriter.java:
```diff
@@ -0,0 +1,177 @@
+    @ProcessElement
+    public void processElement(
+        @Element KV<Integer, Solace.Record> element,
+        @StateId("processing_key") ValueState<Integer> ignoredProcessingKey,
```
The idea is to force parallelization to a maximum number of instances, by controlling the cardinality of the key set in the input. The stateful `DoFn` will force a shuffling, and all elements with the same key will be procesed by the same instance (published by the same client). I guess that it should be the same effect in the stateful `DoFn` without necessarily using this state variable, since there is already a timer. Let me do some tests to make sure the shuffling still happens if I remove the state variable.

iht replied:
Removed in upcoming commit. I have checked and the effect is the same without that state variable. I will resolve now this comment and I will be sending new commits soon (a single commit for this comment).

iht commented:
I have uploaded new commits that should be addressing all the comments that I have resolved.  Each commit should each issue separately, so it would correspond to a single comment, or to a set of some comments that are all about the same issue.

There are still more changes pending, this should not be merged yet, as one of the pending changes is about a bug in the way the Solace responses are handled.


@Abacn @sjvanrossum please review the new commits in the meanwhile, as addressing the remaining changes will take me several days.

iht commented:
Another round of commits just added, each commit should roughly correspond to one of the resolved comments.

Abacn commented:
R: @sjvanrossum

sjvanrossum reviewed (changes requested):
Will look at record callback changes tomorrow.

bzablocki commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -1026,8 +1057,180 @@ abstract static class Builder<T> {
+                      getNumberOfClientsPerWorker(),
+                      getPublishLatencyMetrics()));
+
+      return writer.withOutputTags(FAILED_PUBLISH_TAG, TupleTagList.of(SUCCESSFUL_PUBLISH_TAG));
```
Included in one of the latests commits

sjvanrossum commented on sdks/java/io/csv/src/main/java/org/apache/beam/sdk/io/csv/CsvIO.java:
```diff
@@ -368,8 +368,8 @@ public static Write<Row> writeRows(String to, CSVFormat csvFormat) {
    * // SomeDataClass is a data class configured for Beam to automatically infer its Schema.
-   * @DefaultSchema(AutoValueSchema.class)
-   * @AutoValue
+   * {@literal @}DefaultSchema(AutoValueSchema.class)
+   * {@literal @}AutoValue
```
Remove, not in scope for this PR.

bzablocki replied:
done

sjvanrossum commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/write/UnboundedSolaceWriter.java:
```diff
@@ -165,16 +173,27 @@ public void publishResults(BeamContextWrapper context) {
+                  result,
+                  null,
+                  null,
+                  result.getError() != null
+                      ? checkNotNull(result.getError())
+                      : "SolaceIO.Write: unknown error.");
```
Create a new variable, assign it `result.getError()` and then perform this null check against the variable instead. The reason why error prone keeps complaining about it is because the two calls two `getError` may return two different results, storing the return value is what you want here. The other thing you could do is `Optional.ofNullable(result.getError()).orElse("SolaceIO.Write: unknown error.")`.

bzablocki replied:
Good point, addressed.

sjvanrossum reviewed (commented):
Looking good so far apart from some minor comments to address. :)

bzablocki commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -970,12 +969,14 @@ public Write<T> withSubmissionMode(SubmissionMode submissionMode) {
      *
      * <p>In streaming mode, the publishing latency will be lower, but the throughput will also be
-     * lower.
+     * lower. todo validate the constant sec
```
addressed

bzablocki commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -998,7 +999,12 @@ public Write<T> withSessionServiceFactory(SessionServiceFactory factory) {
+    /** todo docs */
+    public Write<T> withErrorHandler(ErrorHandler<BadRecord, ?> errorHandler) {
+      return toBuilder().setErrorHandler(errorHandler).build();
+    }
```
> is there a place in this class where get* and with* methods are grouped? This seems out of place to me.

I'm not sure if I understand your comment. The get* methods are below grouped together.

bzablocki commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/SolaceIO.java:
```diff
@@ -1079,21 +1089,11 @@ public SolaceOutput expand(PCollection<T> input) {
       PCollection<KV<Integer, Solace.Record>> withShardKeys =
-          withGlobalWindow.apply(
-              "Add shard key", ParDo.of(new AddShardKeyDoFn(getMaxNumOfUsedWorkers())));
+          withGlobalWindow.apply("Add shard key", ParDo.of(new AddShardKeyDoFn(getNumShards())));
```
Agreed. We can add GroupIntoBatches in a separate pull request to avoid delaying this one. This ensures we can proceed without being held up by the GroupIntoBatches implementation, especially since it doesn't affect the public API.

bzablocki commented on sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/BasicAuthJcsmpSessionService.java:
```diff
@@ -119,7 +122,7 @@ public MessageReceiver getReceiver() {
 
   @Override
-  public MessageProducer getProducer(SubmissionMode submissionMode) {
+  public MessageProducer getInitializeProducer(SubmissionMode submissionMode) {
```
done

bzablocki commented:
Hi @Abacn, would you mind merging this PR? It was reviewed in detail by @sjvanrossum :)
